adapters/repos/db/vector/hnsw/delete.go

```diff
@@ -401,11 +402,18 @@ func (h *hnsw) reassignNeighborsOf(deleteList helpers.AllowList, breakCleanUpTom
 	defer h.resetLock.Unlock()
 
 	g, ctx := enterrors.NewErrorGroupWithContextWrapper(h.logger, h.shutdownCtx)
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
 	ch := make(chan uint64)
+	var cancelled atomic.Bool
 
 	for i := 0; i < tombstoneDeletionConcurrency(); i++ {
 		g.Go(func() error {
 			for {
+				if breakCleanUpTombstonedNodes() {
+					cancelled.Store(true)
+					cancel()
```

Comment:
I think that when you have `defer cancel()` in this method then you don't have to explictly call `cancel()` anywhere else in this method

Reply:
I am only invoking the cancel in the places I want to stop the iterations now. Here for example, any thread should be able to detect the change in the context and stop all others.